File: Linked-Lists/removeNthNodeFromEndOfList_method2_leetcode.py
#problem: https://leetcode.com/problems/remove-nth-node-from-end-of-list/ 
# 2-pass approach, using 3 ptrs. - O(2n)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        
        ptr1 = head
        ptr2 = head
        prev = head
        sizeOfLL = 0
        
        while ptr1 != None:         # 1) obtains size of LL
            sizeOfLL+=1
            ptr1 = ptr1.next
        
        if n == sizeOfLL:           # Base case: deleting head from LL
            if head.next != None:   # - if head has A next node
                head = head.next
            else:                   # - if head has no next...
                return None
            
        else:
            nodesToPass = sizeOfLL-n
            while nodesToPass > 0:  # 2) find the node to delete
                prev = ptr2
                ptr2 = ptr2.next
                nodesToPass-=1

            if ptr2.next != None:   # - if another node exists after node-to-be-deleted.
                prev.next = ptr2.next
            else:
                prev.next = None    # - if another node DNE after node-to-be-deleted.
    
        return head

File: Linked-Lists/removeNthNodeFromEndOfList_method2_leetcode.java
// problem: https://leetcode.com/problems/remove-nth-node-from-end-of-list/
// TIME: O(n)
// SPACE: O(1)
// - need to traverse list once for size, then again to find node to remove, so O(2n) -> O(n).

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    public int getListSize(ListNode head){
        ListNode curr = head;
        int size = 0;
        while(curr != null){
            curr = curr.next;
            size++;
        }
        return size;
    }
    
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // - n must be <= size of LL
        
        // want to remove only node in list.
        if((head.next == null && n>0))
            return null;
        
        // get size of LL
        int size = getListSize(head);
        
        // traverse to node to remove.
        // - wherever 'curr' lands, we want to remove from list.
        // - use 'prev' re-route ptrs past 'curr'.
        int numOfNodes = size-n;
        ListNode curr = head;
        ListNode prev = null;
        while(numOfNodes > 0){
            prev = curr;
            curr = curr.next;
            numOfNodes--;
        }
        
        // last node in list is being removed.
        if(curr == null)
            prev.next = null;
        
        // removing head of list
        else if(curr == head)
            head = head.next;
        
        // removing a node in middle of list
        else
            prev.next = curr.next;
        
        return head;
    }
}